File: Average Calculator HTTP Microservice/Average Calculator HTTP Microservice.py
#Develop Average Calculator HTTP Microservice
#Implementation of the Average Calculator microservice using Flask:
from flask import Flask, request, jsonify
import requests
from collections import deque

app = Flask(__name__)

class AverageCalculator:
    def __init__(self, window_size):
        self.window_size = window_size
        self.window_prev_state = deque(maxlen=window_size)
        self.window_curr_state = deque(maxlen=window_size)

    def fetch_numbers(self, number_id):
        if number_id == 'p':
            url = 'http://20.244.56.144/test/primes'
        elif number_id == 'f':
            url = 'http://20.244.56.144/test/fibo'
        elif number_id == 'e':
            url = 'http://20.244.56.144/test/even'
        elif number_id == 'r':
            url = 'http://20.244.56.144/test/random'
        else:
            return jsonify({'error': 'Invalid number ID'}), 400

        try:
            response = requests.get(url, timeout=1)
            response.raise_for_status()
        except requests.RequestException as e:
            return jsonify({'error': 'Failed to fetch numbers'}), 500

        numbers = response.json()['numbers']
        return numbers

    def update_window_state(self, numbers):
        self.window_prev_state = self.window_curr_state
        self.window_curr_state = deque(numbers[:self.window_size], maxlen=self.window_size)

    def calculate_average(self):
        return sum(self.window_curr_state) / len(self.window_curr_state)

@app.route('/numbers/<number_id>', methods=['GET'])
def get_numbers(number_id):
    calculator = AverageCalculator(window_size=10)
    numbers = calculator.fetch_numbers(number_id)
    calculator.update_window_state(numbers)
    avg = calculator.calculate_average()
    return jsonify({
        'windowPrevState': list(calculator.window_prev_state),
        'windowCurrState': list(calculator.window_curr_state),
        'numbers': numbers,
        'avg': avg
    })

if __name__ == '__main__':
    app.run(host='localhost', port=9876)

File: Average Calculator HTTP Microservice.js
//First, let's create a new Node.js project and install the necessary dependencies:
//in bash
//mkdir average-calculator
//cd average-calculator
//npm init -y
//npm install express request

const express = require('express');
const request = require('request');

const app = express();
const PORT = process.env.PORT || 3000;
const WINDOW_SIZE = 10;

let windowPrevState = [];
let windowCurrState = [];

app.get('/numbers/:numberid', (req, res) => {
  const numberId = req.params.numberid;
  const thirdPartyApiUrl = `https://api.example.com/numbers?type=${numberId}`;

  // Fetch numbers from the third-party API
  request(thirdPartyApiUrl, { json: true }, (err, response, body) => {
    if (err || response.statusCode !== 200) {
      return res.status(500).json({ error: 'Error fetching numbers from third-party API' });
    }

    const numbers = body.numbers;

    // Update the window state
    windowPrevState = windowCurrState;
    windowCurrState = [...windowCurrState, ...numbers];

    // Limit the window size
    windowCurrState = windowCurrState.slice(-WINDOW_SIZE);

    // Calculate the average
    const sum = windowCurrState.reduce((acc, num) => acc + num, 0);
    const avg = parseFloat((sum / windowCurrState.length).toFixed(2));

    // Respond with the current window state and the average
    res.json({
      windowPrevState,
      windowCurrState,
      numbers,
      avg
    });
  });
});

app.listen(PORT, () => {
  console.log(`Average calculator microservice listening on port ${PORT}`);
});

//Replace https://api.example.com/numbers?type=${numberId} with the actual third-party API URL and adjust the request options as needed.

//Microservice can be run using node Average Calculator HTTP Microservice.js
//One can test the microservice using a tool like Postman or CURL. For example: curl http://localhost:3000/numbers/e
